File: crypto.js
/**
 * Client-side cryptography module using Web Crypto API
 * Provides AES-GCM encryption with SHA-256 key derivation
 */

class CryptoHelper {
    constructor() {
        this.key = null;
    }
    
    /**
     * Derive encryption key from password using SHA-256
     * @param {string} password - The shared password
     * @returns {Promise<CryptoKey>} - The derived encryption key
     */
    async deriveKey(password) {
        const encoder = new TextEncoder();
        const passwordBuffer = encoder.encode(password);
        
        const hashBuffer = await crypto.subtle.digest('SHA-256', passwordBuffer);
        
        this.key = await crypto.subtle.importKey(
            'raw',
            hashBuffer,
            { name: 'AES-GCM' },
            false,
            ['encrypt', 'decrypt']
        );
        
        return this.key;
    }
    
    /**
     * Encrypt a message using AES-GCM
     * @param {string} message - The plaintext message
     * @returns {Promise<string>} - Base64 encoded encrypted message
     */
    async encrypt(message) {
        if (!this.key) {
            throw new Error('Key not initialized. Call deriveKey first.');
        }
        
        const encoder = new TextEncoder();
        const messageBuffer = encoder.encode(message);
        
        const iv = crypto.getRandomValues(new Uint8Array(12));
        
        const encryptedBuffer = await crypto.subtle.encrypt(
            {
                name: 'AES-GCM',
                iv: iv
            },
            this.key,
            messageBuffer
        );
        
        const encryptedArray = new Uint8Array(encryptedBuffer);
        const combined = new Uint8Array(iv.length + encryptedArray.length);
        combined.set(iv);
        combined.set(encryptedArray, iv.length);
        
        return this.arrayBufferToBase64(combined);
    }
    
    /**
     * Decrypt a message using AES-GCM
     * @param {string} encryptedBase64 - Base64 encoded encrypted message
     * @returns {Promise<string>} - The decrypted plaintext message
     */
    async decrypt(encryptedBase64) {
        if (!this.key) {
            throw new Error('Key not initialized. Call deriveKey first.');
        }
        
        const combined = this.base64ToArrayBuffer(encryptedBase64);
        
        const iv = combined.slice(0, 12);
        const encryptedData = combined.slice(12);
        
        const decryptedBuffer = await crypto.subtle.decrypt(
            {
                name: 'AES-GCM',
                iv: iv
            },
            this.key,
            encryptedData
        );
        
        const decoder = new TextDecoder();
        return decoder.decode(decryptedBuffer);
    }
    
    /**
     * Convert ArrayBuffer to Base64 string
     * @param {Uint8Array} buffer - The buffer to convert
     * @returns {string} - Base64 encoded string
     */
    arrayBufferToBase64(buffer) {
        let binary = '';
        const bytes = new Uint8Array(buffer);
        for (let i = 0; i < bytes.byteLength; i++) {
            binary += String.fromCharCode(bytes[i]);
        }
        return btoa(binary);
    }
    
    /**
     * Convert Base64 string to ArrayBuffer
     * @param {string} base64 - Base64 encoded string
     * @returns {Uint8Array} - The decoded buffer
     */
    base64ToArrayBuffer(base64) {
        const binary = atob(base64);
        const bytes = new Uint8Array(binary.length);
        for (let i = 0; i < binary.length; i++) {
            bytes[i] = binary.charCodeAt(i);
        }
        return bytes;
    }
}

File: crypto.py
"""
Cryptography module for end-to-end encrypted chat.
Provides AES-EAX encryption with SHA-256 key derivation.
"""

import hashlib
from Crypto.Cipher import AES


def get_key(password: str) -> bytes:
    """
    Derive a 256-bit AES key from a password using SHA-256.
    
    Args:
        password: The shared password string
        
    Returns:
        32-byte (256-bit) AES key
    """
    return hashlib.sha256(password.encode('utf-8')).digest()


def encrypt(message: str, key: bytes) -> bytes:
    """
    Encrypt a plaintext message using AES-EAX mode.
    
    Args:
        message: The plaintext message to encrypt
        key: 256-bit AES key
        
    Returns:
        Encrypted message as bytes: nonce + tag + ciphertext
    """
    cipher = AES.new(key, AES.MODE_EAX)
    ciphertext, tag = cipher.encrypt_and_digest(message.encode('utf-8'))
    return cipher.nonce + tag + ciphertext


def decrypt(cipher_bytes: bytes, key: bytes) -> str:
    """
    Decrypt a ciphertext message using AES-EAX mode.
    Verifies message integrity using the authentication tag.
    
    Args:
        cipher_bytes: Encrypted message (nonce + tag + ciphertext)
        key: 256-bit AES key
        
    Returns:
        Decrypted plaintext message
        
    Raises:
        ValueError: If message authentication fails (tampered message)
    """
    nonce = cipher_bytes[:16]
    tag = cipher_bytes[16:32]
    ciphertext = cipher_bytes[32:]
    
    cipher = AES.new(key, AES.MODE_EAX, nonce=nonce)
    plaintext = cipher.decrypt_and_verify(ciphertext, tag)
    
    return plaintext.decode('utf-8')
